### ISSUE TITLE
[Feature Request] C-API: Fallible KernelCompute call for custom operators

### ISSUE THREAD

**Issue by cbourjau** [microsoft/onnxruntime#14287]
### Describe the feature request

Unless I have overlooked something, it seems that the current C-api does not support propagating a failure in the `KernelCompute` call of a custom operator. While some functions in the public c-api return status pointers this is not true for the compute function (and kernel creation). The [relevant c++ code](https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/session/custom_ops.cc#L143-L146) responsible for calling custom operators always returns a `Status::Ok` no matter what. 

~~Unless this functionality already exists somewhere, I am afraid that it would not be possible to extend the existing API in a backwards compatible way. Implementing such error propagation would require a "version 2" of the custom operator API, as far as I can tell. Is this in the realm of possibilities?~~ I think #16479 presents a pretty clean way to implement this feature.

### Describe scenario use case

There are many failable operations. One very simple example is casting `tensor(string)` to `tensor(int)`. Encountering an error during the parsing can only be propagated with a panic which causes the entire program to abort (e.g. The user's Python process crashes rather than normal exception handling). This is unacceptable in most production scenarios and unfortunately significantly diminishes the value of custom operators.

**Comment by xadupre** [comment]
Maybe @wenbingl would know more about this.

**Comment by yuslepukhin** [comment]
The Custom Op API is being changed as we speak. We recommend using C++ for safety. Throwing C++ exceptions out of custom op `Compute` is the recommended method. @adrianlizarraga 

See https://github.com/microsoft/onnxruntime/pull/13427

**Comment by yuslepukhin** [comment]
@cbourjau are you interested in writing custom ops in languages other than C and C++?

**Comment by cbourjau** [comment]
Thanks for the quick response!

Yes, I/we are interested in writing custom operators in Rust. There is already a [prototype](https://github.com/cbourjau/onnxruntime-custom-op) that works great, besides a few shortcomings around the API. The error-propagation is one of those issues and probably the most difficult to resolve given that there is no obvious, forward-compatible way to address it. I am planning to create one or two more issues next week.

About the issue at hand. Rust behaves exactly like C in this regard given that exceptions do not exist in either language. The Rust compiler produces a dynamic library that adhears to the C-ABI. Throwing exceptions requires the use of C++, but there is no standardized and stable way for them to cross the ABI, as I understand it. Every compiler is free to handle exceptions in a different way. Even the same compiler may use them differently based on build configurations. Exceptions may work, but there is no guarantee.

In summary, using C++ exceptions in this scenario seems very scary and unsafe. This is further aggravated by the fact that a common use case of custom operators is that they are compiled independently of the onnxruntime.

Returning an `OrtStatus` rather than throwing an exception is a solution that nicely fits into the C-ABI and is therefore safe.

After writing this I just noticed that something similar is already in the C-API [guidelines](https://github.com/microsoft/onnxruntime/blob/main/docs/C_API_Guidelines.md#4-public-api-that-may-error-out-must-return-ortstatus-pointer-on-error-or-nullptr-on-success)

**Comment by cbourjau** [comment]
I am happy to create a PR with an API proposal that fits the C-API guidelines. Are there further things I should take into consideration? @wenbingl @adrianlizarraga ?

**Comment by yuslepukhin** [comment]
You are welcome to create a PR. The issue may call for a broader design involving many APIs/langs.

**Comment by yuslepukhin** [comment]
Just to let you know, this is still on a radar.
